av1an-cli/src/lib.rs

```diff
@@ -442,164 +442,212 @@ fn confirm(prompt: &str) -> io::Result<bool> {
   }
 }
 
-pub fn parse_cli(args: CliOpts) -> anyhow::Result<EncodeArgs> {
-  let temp = if let Some(path) = args.temp.as_ref() {
-    path.to_str().unwrap().to_owned()
-  } else {
-    format!(".{}", hash_path(args.input.as_path()))
-  };
+/// Given Folder and File paths as inputs
+/// Converts them all to file paths
+/// Converting only depth 1 of Folder paths
+pub fn convert_input(args_paths: Vec<PathBuf>) -> Vec<PathBuf> {
+  let mut validated_inputs: Vec<PathBuf> = Vec::with_capacity(args_paths.len());
+
+  for input_arg in args_paths {
+    if input_arg.is_file() {
+      validated_inputs.push(input_arg.to_path_buf())
+    } else if input_arg.is_dir() {
+      let files: Vec<PathBuf> = std::fs::read_dir(input_arg)
+        .unwrap()
+        .into_iter()
+        .filter_map(Result::ok)
+        .filter_map(|d| {
+          if let Ok(file_type) = d.file_type() {
+            if file_type.is_file() {
+              Some(d.path())
+            } else {
+              None
+            }
+          } else {
+            None
+          }
+        })
+        .collect();
+      validated_inputs.extend(files)
+    }
+  }
 
-  let input = Input::from(args.input.as_path());
+  validated_inputs
+}
 
-  // TODO make an actual constructor for this
-  let mut encode_args = EncodeArgs {
-    frames: 0,
-    log_file: if let Some(log_file) = args.log_file.as_ref() {
-      Path::new(&format!("{}.log", log_file)).to_owned()
-    } else {
-      Path::new(&temp).join("log.log")
-    },
-    ffmpeg_filter_args: if let Some(args) = args.ffmpeg_filter_args.as_ref() {
-      shlex::split(args).ok_or_else(|| anyhow!("Failed to split ffmpeg filter arguments"))?
-    } else {
-      Vec::new()
-    },
-    temp,
-    force: args.force,
-    passes: if let Some(passes) = args.passes {
-      passes
-    } else {
-      args.encoder.get_default_pass()
-    },
-    video_params: if let Some(args) = args.video_params.as_ref() {
-      shlex::split(args).ok_or_else(|| anyhow!("Failed to split video encoder arguments"))?
+/// Returns vector of Encode args ready to be fed to encoder
+pub fn parse_cli(args: CliOpts) -> anyhow::Result<Vec<EncodeArgs>> {
+  let test = convert_input(args.input.clone());
+
+  if test.is_empty() {
+    anyhow!("No valid input file");
```

Comment:
I'm not sure if this actually bubbles up the error, since `anyhow!` just creates the error struct, but does not return it immediately. I think you just need a `return Err(anyhow!( ... ))`.

Reply:
The anyhow crate also has a `bail` macro that does this in one step: `bail!("No valid input file");`